analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/AvoidUnderPosting.cs

```diff
@@ -0,0 +1,144 @@
+
+            compilationStart.RegisterSymbolStartAction(symbolStart =>
+            {
+                var type = (INamedTypeSymbol)symbolStart.Symbol;
+                if (type.IsControllerType())
+                {
+                    symbolStart.RegisterSyntaxNodeAction(nodeContext => ProcessControllerMethods(nodeContext, examinedTypes), SyntaxKind.MethodDeclaration);
+                }
+            }, SymbolKind.NamedType);
+        });
+
+    private static void ProcessControllerMethods(SonarSyntaxNodeReportingContext context, ConcurrentDictionary<ITypeSymbol, bool> examinedTypes)
+    {
+        if (context.SemanticModel.GetDeclaredSymbol(context.Node) is IMethodSymbol method
+            && method.IsControllerMethod())
+        {
+            var modelParameterTypes = method.Parameters
+                .Select(x => x.Type)
+                .OfType<INamedTypeSymbol>()
+                .Where(x => !IgnoreType(x))
+                .SelectMany(x => RelatedTypesToExamine(x))
+                .Distinct();
+            foreach (var modelParameterType in modelParameterTypes)
+            {
+                CheckInvalidProperties(modelParameterType, context, examinedTypes);
+            }
+        }
+    }
+
+    private static void CheckInvalidProperties(INamedTypeSymbol parameterType, SonarSyntaxNodeReportingContext context, ConcurrentDictionary<ITypeSymbol, bool> examinedTypes)
+    {
+        var declaredProperties = new List<IPropertySymbol>();
+        GetAllDeclaredProperties(parameterType, examinedTypes, declaredProperties);
+        var invalidProperties = declaredProperties.Where(x => !CanBeNull(x.Type)
+            && !x.HasAttribute(KnownType.System_ComponentModel_DataAnnotations_RequiredAttribute)
+            && !x.HasAttribute(KnownType.Microsoft_AspNetCore_Mvc_ModelBinding_Validation_ValidateNeverAttribute));
+        foreach (var property in invalidProperties)
+        {
+            var propertySyntax = property.DeclaringSyntaxReferences[0].GetSyntax();
+            context.ReportIssue(Rule, propertySyntax.GetLocation());
+        }
+    }
+
+    private static bool IgnoreType(ITypeSymbol type) =>
+        type.IsAny(IgnoredTypes)
+        || (!type.DeclaringSyntaxReferences.Any()
+            && !type.Is(KnownType.System_Collections_Generic_IEnumerable_T)
+            && !type.Implements(KnownType.System_Collections_Generic_IEnumerable_T));
+
+    private static bool CanBeNull(ITypeSymbol type) =>
+        type is ITypeParameterSymbol { HasValueTypeConstraint: false }
+        || (type.IsReferenceType && type.NullableAnnotation() != NullableAnnotation.NotAnnotated)
+        || type.IsNullableValueType();
+
+    private static void GetAllDeclaredProperties(ITypeSymbol type, ConcurrentDictionary<ITypeSymbol, bool> examinedTypes, List<IPropertySymbol> declaredProperties)
+    {
+        if (type is INamedTypeSymbol namedType && examinedTypes.TryAdd(namedType, true))
```

Comment:
Shouldn't we do the `IgnoreType` check here (because of the recursion)?
```suggestion
        if (type is INamedTypeSymbol namedType && !IgnoreType(namedType) && examinedTypes.TryAdd(namedType, true))
```